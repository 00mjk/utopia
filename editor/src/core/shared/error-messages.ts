export type ErrorMessageSeverity = 'fatal' | 'error' | 'warning'

export type ErrorMessageSource = 'eslint' | 'build' | 'utopia-parser' | 'runtime'

export interface ErrorMessage {
  fileName: string
  startLine: number | null
  startColumn: number | null
  endLine: number | null
  endColumn: number | null
  codeSnippet: string
  severity: ErrorMessageSeverity
  type: string
  message: string
  errorCode: string
  source: ErrorMessageSource
  passTime: number | null
}

export function messageisFatal(e: ErrorMessage): boolean {
  return e.severity === 'fatal'
}

export function messageIsFatalOrError(e: ErrorMessage): boolean {
  return e.severity === 'fatal' || e.severity === 'error'
}

export function messageIsWarning(e: ErrorMessage): boolean {
  return e.severity === 'warning'
}
