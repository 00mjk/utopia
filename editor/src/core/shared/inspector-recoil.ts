import { RecoilRoot, atom, selector, useRecoilState, useRecoilValue } from 'recoil'

export const layoutHoveredState = atom<boolean>({
  key: 'layoutHoveredState', // unique ID (with respect to other atoms/selectors)
  default: false, // default value (aka initial value)
})
