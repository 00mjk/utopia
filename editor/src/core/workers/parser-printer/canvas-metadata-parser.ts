import * as TS from 'typescript'
import {
  AllFlexAlignments,
  AllFlexDirections,
  AllFlexJustifyContents,
  AllFlexWraps,
  FlexAlignment,
  FlexDirection,
  FlexJustifyContent,
  FlexWrap,
  LayoutSystem,
  NormalisedFrame,
} from 'utopia-api'
import {
  descriptionParseError,
  objectKeyParser,
  optionalObjectKeyParser,
  parseAny,
  parseArray,
  parseEnum,
  parseNullable,
  parseNumber,
  parseObject,
  ParseResult,
  parseString,
} from '../../../utils/value-parser-utils'
import {
  convertScenesToUtopiaCanvasComponent,
  createSceneUidFromIndex,
} from '../../model/scene-utils'
import {
  applicative2Either,
  applicative4Either,
  applicative5Either,
  applicative8Either,
  Either,
  foldEither,
  left,
  right,
} from '../../shared/either'
import { UtopiaJSXComponent } from '../../shared/element-template'
import {
  CanvasElementMetadataMap,
  CanvasMetadata,
  CanvasMetadataRightBeforePrinting,
  ElementCanvasMetadata,
  isParsedJSONFailure,
  ParsedJSONResult,
  PrintedCanvasMetadata,
  SceneContainer,
  SceneFlexContainer,
  SceneMetadata,
} from '../../shared/project-file-types'
import { parsedJSONFailure, parsedJSONSuccess } from '../common/project-file-utils'
import { createCodeSnippet } from '../ts/ts-utils'

function parseFlexDirection(value: unknown): ParseResult<FlexDirection> {
  return parseEnum(AllFlexDirections)(value)
}

function parseFlexAlignment(value: unknown): ParseResult<FlexAlignment> {
  return parseEnum(AllFlexAlignments)(value)
}

function parseFlexJustifyContent(value: unknown): ParseResult<FlexJustifyContent> {
  return parseEnum(AllFlexJustifyContents)(value)
}

function parseFlexWrap(value: unknown): ParseResult<FlexWrap> {
  return parseEnum(AllFlexWraps)(value)
}

function parseNormalisedFrame(value: unknown): ParseResult<NormalisedFrame> {
  return applicative4Either(
    (frameLeft, frameTop, frameWidth, frameHeight) => {
      return {
        left: frameLeft,
        top: frameTop,
        width: frameWidth,
        height: frameHeight,
      }
    },
    objectKeyParser(parseNumber, 'left')(value),
    objectKeyParser(parseNumber, 'top')(value),
    objectKeyParser(parseNumber, 'width')(value),
    objectKeyParser(parseNumber, 'height')(value),
  )
}

function parseSceneFlexContainer(value: unknown): ParseResult<SceneFlexContainer> {
  return applicative2Either(
    (otherPropsApplicative, paddingApplicative) =>
      ({
        ...otherPropsApplicative,
        ...paddingApplicative,
      } as SceneFlexContainer),
    applicative4Either(
      (paddingTop, paddingRight, paddingBottom, paddingLeft) => {
        let sceneFlexContainer: Partial<SceneFlexContainer> = {}
        if (paddingTop != null) {
          sceneFlexContainer.paddingTop = paddingTop
        }
        if (paddingRight != null) {
          sceneFlexContainer.paddingRight = paddingRight
        }
        if (paddingBottom != null) {
          sceneFlexContainer.paddingBottom = paddingBottom
        }
        if (paddingLeft != null) {
          sceneFlexContainer.paddingLeft = paddingLeft
        }
        return sceneFlexContainer
      },
      objectKeyParser(parseNullable(parseNumber), 'paddingTop')(value),
      objectKeyParser(parseNullable(parseNumber), 'paddingRight')(value),
      objectKeyParser(parseNullable(parseNumber), 'paddingBottom')(value),
      objectKeyParser(parseNullable(parseNumber), 'paddingLeft')(value),
    ),
    applicative8Either(
      (
        layoutSystem,
        flexDirection,
        alignContent,
        alignItems,
        justifyContent,
        flexWrap,
        gapCross,
        gapMain,
      ) => {
        let sceneFlexContainer: SceneFlexContainer = {
          layoutSystem,
        }
        if (flexDirection != null) {
          sceneFlexContainer.flexDirection = flexDirection
        }
        if (alignContent != null) {
          sceneFlexContainer.alignContent = alignContent
        }
        if (alignItems != null) {
          sceneFlexContainer.alignItems = alignItems
        }
        if (justifyContent != null) {
          sceneFlexContainer.justifyContent = justifyContent
        }
        if (flexWrap != null) {
          sceneFlexContainer.flexWrap = flexWrap
        }
        if (gapCross != null) {
          sceneFlexContainer.gapCross = gapCross
        }
        if (gapMain != null) {
          sceneFlexContainer.gapMain = gapMain
        }
        return sceneFlexContainer
      },
      objectKeyParser(parseEnum([LayoutSystem.Flex]), 'layoutSystem')(value),
      objectKeyParser(parseNullable(parseFlexDirection), 'flexDirection')(value),
      objectKeyParser(parseNullable(parseFlexAlignment), 'alignContent')(value),
      objectKeyParser(parseNullable(parseFlexAlignment), 'alignItems')(value),
      objectKeyParser(parseNullable(parseFlexJustifyContent), 'justifyContent')(value),
      objectKeyParser(parseNullable(parseFlexWrap), 'wrap')(value),
      objectKeyParser(parseNullable(parseNumber), 'gapCross')(value),
      objectKeyParser(parseNullable(parseNumber), 'gapMain')(value),
    ),
  )
}

function parseSceneContainer(value: unknown): ParseResult<SceneContainer> {
  if (typeof value === 'object' && !Array.isArray(value) && value != null) {
    const valueAsObject: any = value
    switch (valueAsObject.layoutSystem) {
      case LayoutSystem.PinSystem:
        return right({
          layoutSystem: LayoutSystem.PinSystem,
        })
      case LayoutSystem.Flex:
        return parseSceneFlexContainer(value)
      default:
        return left(descriptionParseError('Scene container has an invalid layout type.'))
    }
  } else {
    return left(descriptionParseError('Scene container is not an object.'))
  }
}

function parseSceneMetadata(value: unknown, index: number): ParseResult<SceneMetadata> {
  return applicative5Either(
    (component, props, frame, container, label) => {
      var sceneMetadata: SceneMetadata = {
        uid: createSceneUidFromIndex(index),
        component: component,
        props: props,
        frame: frame,
        container: container,
      }
      if (label != null) {
        sceneMetadata.label = label
      }
      return sceneMetadata
    },
    objectKeyParser(parseNullable(parseString), 'component')(value),
    objectKeyParser(parseObject(parseAny), 'props')(value),
    objectKeyParser(parseNormalisedFrame, 'frame')(value),
    objectKeyParser(parseSceneContainer, 'container')(value),
    optionalObjectKeyParser(parseString, 'label')(value),
  )
}

function parseElementCanvasMetadata(value: unknown): ParseResult<ElementCanvasMetadata> {
  if (typeof value === 'object' && !Array.isArray(value) && value != null) {
    return right({})
  } else {
    return left(descriptionParseError('Canvas metadata is not an object.'))
  }
}

function parseCanvasElementMetadataMap(value: unknown): ParseResult<CanvasElementMetadataMap> {
  return parseObject(parseElementCanvasMetadata)(value)
}

export function parsePrintedCanvasMetadata(value: unknown): ParseResult<PrintedCanvasMetadata> {
  return applicative2Either(
    (scenes, elementMetadata) => {
      return {
        scenes: scenes,
        elementMetadata: elementMetadata,
      }
    },
    objectKeyParser(parseNullable(parseArray(parseSceneMetadata)), 'scenes')(value),
    objectKeyParser(parseCanvasElementMetadataMap, 'elementMetadata')(value),
  )
}

export function jsonToExpression(json: any): TS.Expression {
  switch (typeof json) {
    case 'object':
      if (json == null) {
        return TS.createNull()
      }
      if (Array.isArray(json)) {
        const arrayExpressions: Array<TS.Expression> = json.map(jsonToExpression)
        return TS.createArrayLiteral(arrayExpressions)
      } else {
        const keys = Object.keys(json)
        const objectPropertyExpressions: Array<TS.ObjectLiteralElementLike> = keys.map((key) => {
          return TS.createPropertyAssignment(
            TS.createStringLiteral(key),
            jsonToExpression(json[key]),
          )
        })
        return TS.createObjectLiteral(objectPropertyExpressions)
      }
    case 'string':
      return TS.createStringLiteral(json)
    case 'boolean':
      return json ? TS.createTrue() : TS.createFalse()
    case 'number':
      if (json < 0) {
        return TS.createPrefix(TS.SyntaxKind.MinusToken, TS.createLiteral(json * -1))
      } else {
        return TS.createLiteral(json)
      }
    case 'undefined':
      return TS.createIdentifier('undefined')
    default:
      throw new Error(`Unsure how to handle ${JSON.stringify(json)}`)
  }
}

export function canvasMetadataToExpression(
  canvasMetadata: CanvasMetadataRightBeforePrinting,
): TS.Expression {
  return jsonToExpression(canvasMetadata)
}

export function parseCanvasMetadata(sourceFile: TS.SourceFile, node: TS.Node): ParsedJSONResult {
  const unprocessedCanvasMetadata = parseJSON(sourceFile, node)
  if (unprocessedCanvasMetadata.type === 'FAILURE') {
    return unprocessedCanvasMetadata
  } else {
    const canvasMetadata: CanvasMetadata = unprocessedCanvasMetadata.value
    return parsedJSONSuccess(canvasMetadata)
  }
}

export function convertPrintedMetadataToCanvasMetadata(
  metadata: Either<unknown, PrintedCanvasMetadata>,
): {
  sanitizedCanvasMetadata: Either<unknown, CanvasMetadata>
  utopiaComponentFromSceneMetadata: UtopiaJSXComponent | null
} {
  return foldEither(
    (l) => ({ sanitizedCanvasMetadata: left(l), utopiaComponentFromSceneMetadata: null }),
    (printedMetadata: PrintedCanvasMetadata) => ({
      sanitizedCanvasMetadata: right({}),
      utopiaComponentFromSceneMetadata: convertScenesToUtopiaCanvasComponent(
        printedMetadata.scenes,
      ),
    }),
    metadata,
  )
}

function parseJSONObject(
  sourceFile: TS.SourceFile,
  node: TS.ObjectLiteralExpression,
): ParsedJSONResult {
  let result: { [key: string]: any } = {}
  for (const property of node.properties) {
    if (TS.isPropertyAssignment(property)) {
      const propertyResult = parseJSON(sourceFile, property.initializer)
      if (isParsedJSONFailure(propertyResult)) {
        return propertyResult
      } else {
        const propertyName = TS.isStringLiteral(property.name)
          ? property.name.text
          : property.name.getText(sourceFile)
        result[propertyName] = propertyResult.value
      }
    } else {
      return createJsonError(sourceFile, node, 'expected property assignment')
    }
  }
  return parsedJSONSuccess(result)
}

function parseJSONArray(
  sourceFile: TS.SourceFile,
  node: TS.ArrayLiteralExpression,
): ParsedJSONResult {
  let result: Array<any> = []
  for (const child of node.elements) {
    const childResult = parseJSON(sourceFile, child)
    if (isParsedJSONFailure(childResult)) {
      return childResult
    } else {
      result.push(childResult.value)
    }
  }
  return parsedJSONSuccess(result)
}

function parsePrefixUnaryExpression(
  sourceFile: TS.SourceFile,
  node: TS.PrefixUnaryExpression,
): ParsedJSONResult {
  // Cater for negative numbers, because of course they're done in a weird way.
  const operator = node.operator
  if (operator === TS.SyntaxKind.MinusToken) {
    const operand = node.operand
    if (TS.isNumericLiteral(operand)) {
      return parsedJSONSuccess(Number.parseFloat(operand.getText(sourceFile)) * -1)
    }
  }
  return createJsonError(sourceFile, node, `expected '-' prefix for negative number`)
}

function parseJSON(sourceFile: TS.SourceFile, node: TS.Node): ParsedJSONResult {
  if (TS.isObjectLiteralExpression(node)) {
    return parseJSONObject(sourceFile, node)
  } else if (TS.isArrayLiteralExpression(node)) {
    return parseJSONArray(sourceFile, node)
  } else if (TS.isStringLiteral(node)) {
    return parsedJSONSuccess(node.text)
  } else if (TS.isNumericLiteral(node)) {
    return parsedJSONSuccess(Number.parseFloat(node.getText(sourceFile)))
  } else if (TS.isPrefixUnaryExpression(node)) {
    return parsePrefixUnaryExpression(sourceFile, node)
  } else {
    switch (node.kind) {
      case TS.SyntaxKind.TrueKeyword:
        return parsedJSONSuccess(true)
      case TS.SyntaxKind.FalseKeyword:
        return parsedJSONSuccess(false)
      case TS.SyntaxKind.NullKeyword:
        return parsedJSONSuccess(null)
      default:
        return createJsonError(sourceFile, node, 'invalid content for JSON')
    }
  }
}

function createJsonError(sourceFile: TS.SourceFile, node: TS.Node, reason: string) {
  const nodeStart = node.getStart(sourceFile)
  const nodeEnd = node.getEnd()
  const codeSnippet = createCodeSnippet(sourceFile, nodeStart, nodeEnd)
  let { line: startLine, character: startCol } = sourceFile.getLineAndCharacterOfPosition(nodeStart)
  let { line: endLine, character: endCol } = sourceFile.getLineAndCharacterOfPosition(nodeEnd)

  return parsedJSONFailure(
    node,
    codeSnippet,
    reason,
    startLine + 1, // typescript lines and cols are zero-based, correcting it to be one-based
    startCol + 1,
    endLine + 1,
    endCol + 1,
  )
}
