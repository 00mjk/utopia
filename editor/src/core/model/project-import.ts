import * as JSZip from 'jszip'
import { JSZipObject } from 'jszip'
import { isText } from 'istextorbinary'
import { ProjectContents, RevisionsState } from '../shared/project-file-types'
import {
  assetFile,
  codeFile,
  directory,
  fileTypeFromFileName,
  imageFile,
  uiJsFile,
} from './project-file-utils'
import { lintAndParse } from '../workers/parser-printer/parser-printer'
import { assetResultForBase64, getFileExtension, imageResultForBase64 } from '../shared/file-utils'

async function attemptedTextFileLoad(fileName: string, file: JSZipObject): Promise<string | null> {
  const fileBuffer = await file.async('nodebuffer')
  if (isText(fileName, fileBuffer)) {
    return file.async('string')
  } else {
    return null
  }
}

export async function importZippedGitProject(zipped: JSZip): Promise<ProjectContents> {
  let loadedProject: ProjectContents = {}
  let promises: Array<Promise<void>> = []

  const loadFile = async (fileName: string, file: JSZip.JSZipObject) => {
    const shiftedFileName = fileName.replace(/[^\/]*\//, '/') // Github uses the project name and a commit hash as the root directory
    if (file.dir) {
      loadedProject[shiftedFileName] = directory()
    } else {
      const expectedFileType = fileTypeFromFileName(shiftedFileName)
      switch (expectedFileType) {
        case 'ASSET_FILE': {
          const fileType = getFileExtension(shiftedFileName)
          const base64 = await file.async('base64')
          const assetResult = assetResultForBase64(shiftedFileName, base64)
          loadedProject[shiftedFileName] = assetFile()
          // TODO trigger upload
          break
        }
        case 'IMAGE_FILE': {
          const fileType = getFileExtension(shiftedFileName)
          const base64 = await file.async('base64')
          const imageResult = await imageResultForBase64(shiftedFileName, fileType, base64)
          loadedProject[shiftedFileName] = imageFile(
            fileType,
            imageResult.dataUrl,
            imageResult.size.width,
            imageResult.size.height,
            imageResult.hash,
          )
          // TODO trigger upload
          break
        }
        case 'UI_JS_FILE':
        case 'CODE_FILE':
          const loadedFile = await attemptedTextFileLoad(shiftedFileName, file)
          if (loadedFile == null) {
            // FIXME Client should show an error if loading a text file fails
            console.error(`Failed to parse file ${shiftedFileName} as a text file`)
          } else {
            if (expectedFileType === 'UI_JS_FILE') {
              const parsedUIFile = lintAndParse(shiftedFileName, loadedFile)
              loadedProject[shiftedFileName] = uiJsFile(
                parsedUIFile,
                null,
                RevisionsState.BothMatch,
                Date.now(),
              )
            } else {
              loadedProject[shiftedFileName] = codeFile(loadedFile, null)
            }
          }
          break
        default:
          const _exhaustiveCheck: never = expectedFileType
          throw new Error(`Unknown file type ${expectedFileType}`)
      }
    }
  }

  zipped.forEach((fileName, file) => {
    promises.push(loadFile(fileName, file))
  })

  await Promise.all(promises)
  return loadedProject
}
