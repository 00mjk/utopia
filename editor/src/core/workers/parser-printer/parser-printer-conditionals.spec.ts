import { applyPrettier } from 'utopia-vscode-common'
import { BakedInStoryboardUID } from '../../model/scene-utils'
import { foldParsedTextFile } from '../../shared/project-file-types'
import { clearParseResultUniqueIDsAndEmptyBlocks, testParseCode } from './parser-printer.test-utils'

describe('JSX Parser', () => {
  it('ternaries with basic values', () => {
    const code = applyPrettier(
      `
/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >{false ? 'cake' : 'cookie'}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'${BakedInStoryboardUID}'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid={'scene-aaa'}
    />
  </Storyboard>
)
`,
      false,
    ).formatted
    const parseResult = clearParseResultUniqueIDsAndEmptyBlocks(testParseCode(code))
    foldParsedTextFile(
      (failure) => {
        fail(failure)
      },
      (success) => {
        expect(success.imports).toMatchSnapshot()
        expect(success.topLevelElements).toMatchSnapshot()
      },
      (unparsed) => {
        fail(unparsed)
      },
      parseResult,
    )
  })
  it('ternaries with elements', () => {
    const code = applyPrettier(
      `
/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >{false ? <div /> : <span />}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'${BakedInStoryboardUID}'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid={'scene-aaa'}
    />
  </Storyboard>
)
`,
      false,
    ).formatted
    const parseResult = clearParseResultUniqueIDsAndEmptyBlocks(testParseCode(code))
    foldParsedTextFile(
      (failure) => {
        fail(failure)
      },
      (success) => {
        expect(success.imports).toMatchSnapshot()
        expect(success.topLevelElements).toMatchSnapshot()
      },
      (unparsed) => {
        fail(unparsed)
      },
      parseResult,
    )
  })
})
