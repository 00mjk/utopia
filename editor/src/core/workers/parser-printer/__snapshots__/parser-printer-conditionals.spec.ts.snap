// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX Parser ternaries with basic values 1`] = `
Object {
  "react": Object {
    "importedAs": "React",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
  "utopia-api": Object {
    "importedAs": null,
    "importedFromWithin": Array [
      Object {
        "alias": "Scene",
        "name": "Scene",
      },
      Object {
        "alias": "Storyboard",
        "name": "Storyboard",
      },
      Object {
        "alias": "jsx",
        "name": "jsx",
      },
    ],
    "importedWithName": null,
  },
}
`;

exports[`JSX Parser ternaries with basic values 2`] = `
Array [
  Object {
    "arbitraryJSBlock": null,
    "isFunction": true,
    "name": "App",
    "param": Object {
      "boundParam": Object {
        "defaultExpression": null,
        "paramName": "props",
        "type": "REGULAR_PARAM",
      },
      "dotDotDotToken": false,
      "type": "PARAM",
    },
    "propsUsed": Array [],
    "rootElement": Object {
      "children": Array [
        Object {
          "text": "
      ",
          "type": "JSX_TEXT_BLOCK",
          "uniqueID": "",
        },
        Object {
          "condition": Object {
            "type": "ATTRIBUTE_VALUE",
            "value": false,
          },
          "type": "JSX_CONDITIONAL_EXPRESSION",
          "uniqueID": "",
          "whenFalse": Object {
            "type": "ATTRIBUTE_VALUE",
            "value": "cookie",
          },
          "whenTrue": Object {
            "type": "ATTRIBUTE_VALUE",
            "value": "cake",
          },
        },
        Object {
          "text": "
    ",
          "type": "JSX_TEXT_BLOCK",
          "uniqueID": "",
        },
      ],
      "name": Object {
        "baseVariable": "div",
        "propertyPath": Object {
          "propertyElements": Array [],
        },
      },
      "props": Object {
        "data-uid": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": "318",
        },
        "layout": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "layoutSystem": "pinSystem",
          },
        },
        "style": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "width": "100%",
          },
        },
      },
      "type": "JSX_ELEMENT",
    },
    "type": "UTOPIA_JSX_COMPONENT",
  },
  Object {
    "arbitraryJSBlock": null,
    "isFunction": false,
    "name": "storyboard",
    "param": null,
    "propsUsed": Array [],
    "rootElement": Object {
      "children": Array [
        Object {
          "children": Array [],
          "name": Object {
            "baseVariable": "Scene",
            "propertyPath": Object {
              "propertyElements": Array [],
            },
          },
          "props": Object {
            "component": Object {
              "definedElsewhere": Array [
                "App",
              ],
              "javascript": "App",
              "sourceMap": Object {
                "file": "code.tsx",
                "mappings": "OAiBkBA",
                "names": Array [
                  "App",
                ],
                "sources": Array [
                  "code.tsx",
                ],
                "sourcesContent": Array [
                  "/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >
      {false ? 'cake' : 'cookie'}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'utopia-storyboard-uid'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
      component={App}
      layout={{ layoutSystem: 'pinSystem' }}
      props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
      data-uid={'scene-aaa'}
    />
  </Storyboard>
)
",
                ],
                "version": 3,
              },
              "transpiledJavascript": "return App;",
              "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
              "uniqueID": "",
            },
            "data-uid": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": "scene-aaa",
            },
            "layout": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "layoutSystem": "pinSystem",
              },
            },
            "props": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "layout": Object {
                  "bottom": 0,
                  "left": 0,
                  "right": 0,
                  "top": 0,
                },
              },
            },
            "style": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "height": 812,
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": 375,
              },
            },
          },
          "type": "JSX_ELEMENT",
        },
      ],
      "name": Object {
        "baseVariable": "Storyboard",
        "propertyPath": Object {
          "propertyElements": Array [],
        },
      },
      "props": Object {
        "data-uid": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": "utopia-storyboard-uid",
        },
        "layout": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "layoutSystem": "pinSystem",
          },
        },
      },
      "type": "JSX_ELEMENT",
    },
    "type": "UTOPIA_JSX_COMPONENT",
  },
]
`;

exports[`JSX Parser ternaries with elements 1`] = `
Object {
  "react": Object {
    "importedAs": "React",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
  "utopia-api": Object {
    "importedAs": null,
    "importedFromWithin": Array [
      Object {
        "alias": "Scene",
        "name": "Scene",
      },
      Object {
        "alias": "Storyboard",
        "name": "Storyboard",
      },
      Object {
        "alias": "jsx",
        "name": "jsx",
      },
    ],
    "importedWithName": null,
  },
}
`;

exports[`JSX Parser ternaries with elements 2`] = `
Array [
  Object {
    "arbitraryJSBlock": null,
    "isFunction": true,
    "name": "App",
    "param": Object {
      "boundParam": Object {
        "defaultExpression": null,
        "paramName": "props",
        "type": "REGULAR_PARAM",
      },
      "dotDotDotToken": false,
      "type": "PARAM",
    },
    "propsUsed": Array [],
    "rootElement": Object {
      "children": Array [
        Object {
          "text": "
      ",
          "type": "JSX_TEXT_BLOCK",
          "uniqueID": "",
        },
        Object {
          "condition": Object {
            "type": "ATTRIBUTE_VALUE",
            "value": false,
          },
          "type": "JSX_CONDITIONAL_EXPRESSION",
          "uniqueID": "",
          "whenFalse": Object {
            "definedElsewhere": Array [
              "React",
            ],
            "javascript": "<span />",
            "sourceMap": Object {
              "file": "code.tsx",
              "mappings": "OAS2BA,iCAAFC",
              "names": Array [
                "span",
                " ",
              ],
              "sources": Array [
                "code.tsx",
              ],
              "sourcesContent": Array [
                "/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >
      {false ? <div /> : <span />}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'utopia-storyboard-uid'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
      component={App}
      layout={{ layoutSystem: 'pinSystem' }}
      props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
      data-uid={'scene-aaa'}
    />
  </Storyboard>
)
",
              ],
              "version": 3,
            },
            "transpiledJavascript": "return React.createElement(\\"span\\", null);",
            "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
            "uniqueID": "",
          },
          "whenTrue": Object {
            "definedElsewhere": Array [
              "React",
            ],
            "javascript": "<div />",
            "sourceMap": Object {
              "file": "code.tsx",
              "mappings": "OASiBA,gCAAFC",
              "names": Array [
                "div",
                " ",
              ],
              "sources": Array [
                "code.tsx",
              ],
              "sourcesContent": Array [
                "/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >
      {false ? <div /> : <span />}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'utopia-storyboard-uid'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
      component={App}
      layout={{ layoutSystem: 'pinSystem' }}
      props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
      data-uid={'scene-aaa'}
    />
  </Storyboard>
)
",
              ],
              "version": 3,
            },
            "transpiledJavascript": "return React.createElement(\\"div\\", null);",
            "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
            "uniqueID": "",
          },
        },
        Object {
          "text": "
    ",
          "type": "JSX_TEXT_BLOCK",
          "uniqueID": "",
        },
      ],
      "name": Object {
        "baseVariable": "div",
        "propertyPath": Object {
          "propertyElements": Array [],
        },
      },
      "props": Object {
        "data-uid": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": "318",
        },
        "layout": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "layoutSystem": "pinSystem",
          },
        },
        "style": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "backgroundColor": "#FFFFFF",
            "height": "100%",
            "width": "100%",
          },
        },
      },
      "type": "JSX_ELEMENT",
    },
    "type": "UTOPIA_JSX_COMPONENT",
  },
  Object {
    "arbitraryJSBlock": null,
    "isFunction": false,
    "name": "storyboard",
    "param": null,
    "propsUsed": Array [],
    "rootElement": Object {
      "children": Array [
        Object {
          "children": Array [],
          "name": Object {
            "baseVariable": "Scene",
            "propertyPath": Object {
              "propertyElements": Array [],
            },
          },
          "props": Object {
            "component": Object {
              "definedElsewhere": Array [
                "App",
              ],
              "javascript": "App",
              "sourceMap": Object {
                "file": "code.tsx",
                "mappings": "OAiBkBA",
                "names": Array [
                  "App",
                ],
                "sources": Array [
                  "code.tsx",
                ],
                "sourcesContent": Array [
                  "/** @jsx jsx */
import * as React from 'react'
import { Scene, Storyboard, jsx } from 'utopia-api'
export var App = (props) => {
  return (
    <div
      style={{ width: '100%', height: '100%', backgroundColor: '#FFFFFF' }}
      layout={{ layoutSystem: 'pinSystem' }}
    >
      {false ? <div /> : <span />}
    </div>
  )
}
export var storyboard = (
  <Storyboard data-uid={'utopia-storyboard-uid'} layout={{ layoutSystem: 'pinSystem' }}>
    <Scene
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
      component={App}
      layout={{ layoutSystem: 'pinSystem' }}
      props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
      data-uid={'scene-aaa'}
    />
  </Storyboard>
)
",
                ],
                "version": 3,
              },
              "transpiledJavascript": "return App;",
              "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
              "uniqueID": "",
            },
            "data-uid": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": "scene-aaa",
            },
            "layout": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "layoutSystem": "pinSystem",
              },
            },
            "props": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "layout": Object {
                  "bottom": 0,
                  "left": 0,
                  "right": 0,
                  "top": 0,
                },
              },
            },
            "style": Object {
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "height": 812,
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": 375,
              },
            },
          },
          "type": "JSX_ELEMENT",
        },
      ],
      "name": Object {
        "baseVariable": "Storyboard",
        "propertyPath": Object {
          "propertyElements": Array [],
        },
      },
      "props": Object {
        "data-uid": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": "utopia-storyboard-uid",
        },
        "layout": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": Object {
            "layoutSystem": "pinSystem",
          },
        },
      },
      "type": "JSX_ELEMENT",
    },
    "type": "UTOPIA_JSX_COMPONENT",
  },
]
`;
